experimental/errorsource/error_source_middleware.go

```diff
@@ -54,13 +55,28 @@ func DownstreamError(err error, override bool) error {
 
 // SourceError returns an error with the source
 // If source is already defined, it will return it, or you can override
-func SourceError(source backend.ErrorSource, err error, override bool) error {
+func SourceError(source backend.ErrorSource, err error, override bool) Error {
 	var sourceError Error
 	if errors.As(err, &sourceError) && !override {
-		return err // already has a source
+		return sourceError // already has a source
 	}
 	return Error{
 		Source: source,
 		Err:    err,
 	}
 }
+
+// Response returns an error DataResponse given status, source of the error and message.
+func Response(err error) backend.DataResponse {
+	e := SourceError(backend.ErrorSourcePlugin, err, false)
+	return backend.DataResponse{
+		Error:       e,
+		ErrorSource: e.Source,
+		Status:      e.Status,
+	}
```

Comment:
I found this part a bit confusing, especially the first line. I think it's easier to follow if we just unwrap the `errorsource.Error` using `errors.As`, and we use the data inside it instead:

```suggestion
	var e Error
	if !errors.As(err, &e) {
		// generic error, default to "plugin" error source
		return backend.DataResponse{
			Error:       err,
			ErrorSource: backend.ErrorSourcePlugin,
			Status:      backend.Status(0), // alternatively, `backend.StatusUnknown` which is 500
		}
	}
	return backend.DataResponse{
		Error:       err,
		ErrorSource: e.Source,
		Status:      e.Status,
	}
```